### PR TITLE
fix: move SASjsRequest array from BaseJobExecutor class to RequestClient class

### LINKED ISSUE #532: Move SASjsRequest array from BaseJobExecutor class to RequestClient class
We want to store all requests in an array. Currently we are doing so by maintaining an array in `BaseJobExecutor` class. But there is an issue in this way. We are not able to access this array from `FileUploader`. So, we decided to have a singleton class which would store these requests. And luckily we already have a class `RequestClient` that is  behaving as singleton.
Also it makes complete sense to have the requests array in `RequestClient` so that any request to the server (coming from anywhere in the codebase) can be stored there.

### PR DESCRIPTION
**Pull request by sabhas**
## Issue

closes #532 

## Intent

- append SASjsRequest from fileUpload
- throw error, if null or undefined passed to getValidJson
- set debug: false in config of  fileUploader tests

## Implementation

- added SASjsRequest array in RequestClient class
- moved appendRequest, getSasRequests and getSasRequests from BaseJobExecutor to RequestClient
- updated getSasRequests method in main SASjs.ts file to get SASjsRequests from RequestClient
- added check for null or undefined in getValidJson function

## Checks

No PR (that involves a non-trivial code change) should be merged, unless all items below are confirmed!  If an urgent fix is needed - use a tar file.

- [x] All `sasjs-cli` unit tests are passing (`npm test`).
Mocked:
image
Server:
image
- [x] All `sasjs-tests` are passing (instructions available [here](https://github.com/sasjs/adapter/blob/master/sasjs-tests/README.md)).
VIYA:
image
SAS9:
image
- [x] [Data Controller](https://datacontroller.io) builds and is functional on both SAS 9 and Viya
VIYA:
image
SAS9:
image

### CODE REVIEW

File: src/SASViyaApiClient.ts [hunk 1]
```diff
@@ -51,6 +51,10 @@ export class SASViyaApiClient {
   )
   private folderMap = new Map<string, Job[]>()
 
+  public appendRequest(response: any, program: string, debug: boolean) {
```
Comment (allanbowe) [inline src/SASViyaApiClient.ts:54]: a JSDoc header here would be great

File: src/request/RequestClient.ts [hunk 2]
```diff
@@ -83,6 +86,53 @@ export class RequestClient implements HttpClient {
     return this.httpClient.defaults.baseURL || ''
   }
 
+  public getRequests = () => this.requests
+
+  public clearRequests = () => {
+    this.requests = []
+  }
+
+  public appendRequest(response: any, program: string, debug: boolean) {
```
Comment (allanbowe) [inline src/request/RequestClient.ts:95]: JSDoc header here also

**Review by allanbowe — COMMENTED** [review]
Would be nice to see:
* JSDoc headers for all public functions/methods
* Some comments explaining the intent behind the code

We should also ensure this PR is well tested on [Data Controller](https://datacontroller.io) before merging

File: src/SASViyaApiClient.ts [hunk 3]
```diff
@@ -51,6 +51,16 @@ export class SASViyaApiClient {
   )
   private folderMap = new Map<string, Job[]>()
 
+  /**
+   * A helper method used to call appendRequest method of RequestClient
+   * @param response - response from sasjs request
+   * @param program - name of program
+   * @param debug - a boolean that indicates whether debug was enabled or not
+   */
+  public appendRequest(response: any, program: string, debug: boolean) {
```
Comment (sabhas) [inline src/SASViyaApiClient.ts:60]: @YuryShkoda , I think name of this helper method should be changed as `RequestClient` class has a method with same name too. Also this method if the helper of that method. At some places we are appending `SASjsRequest` array but don't have access to `RequestClient` instance. Instead we have access to instance of `SASViyaApiClient` and this `SASViyaApiClient` instance have access to `RequestClient`. Therefore, I introduced this helper method.

Comment (sabhas) [reply src/SASViyaApiClient.ts:60]: What should I name this method?

File: src/FileUploader.ts [hunk 4]
```diff
@@ -66,6 +66,7 @@ export class FileUploader {
     return this.requestClient
       .post(uploadUrl, formData, undefined, 'application/json', headers)
       .then(async (res) => {
+        this.requestClient!.appendRequest(res, sasJob, this.sasjsConfig.debug)
```
Comment (krishna-acondy) [inline src/FileUploader.ts:69]: I think it'd be best if we didn't have to ask the request client to append the request - it should be able to do it internally whenever get/post/delete/patch/put are called.

**Review by krishna-acondy — CHANGES_REQUESTED** [review]

File: src/FileUploader.ts (hunk 4 above)
Comment (allanbowe) [reply src/FileUploader.ts:69]: As I understand it, there are a few issues with this suggestion:
* Which request to attach it to (some SAS calls comprise several network requests)
* How to capture the necessary metadata (program path, debug state, log, response, etc)

**Comment by VladislavParhomchik** [comment]
sasjs-tests failedimage

**Review by VladislavParhomchik — DISMISSED** [review]

File: src/SASjs.ts [hunk 5]
```diff
@@ -914,10 +912,12 @@ export default class SASjs {
       this.sasjsConfig.serverUrl = this.sasjsConfig.serverUrl.slice(0, -1)
     }
 
-    this.requestClient = new RequestClient(
-      this.sasjsConfig.serverUrl,
-      this.sasjsConfig.allowInsecureRequests
-    )
+    if (!this.requestClient) {
+      this.requestClient = new RequestClient(
```
Comment (krishna-acondy) [inline src/SASjs.ts:916]: If the request client has already been instantiated, we should update the serverUrl and allowInsecureRequests parameters in the existing instance.

**Review by krishna-acondy — CHANGES_REQUESTED** [review]

**Review by krishna-acondy — APPROVED** [review]

**Comment by medjedovicm** [comment]
Windows CLI tests passing:
image

**Comment by medjedovicm** [comment]
SASJS-TESTS:
VIYA
viyatests

SAS9
sas9tests

**Pull request merged**